github_workflow.md

```diff
@@ -0,0 +1,153 @@
+# GitHub Milestone and Issue Workflow
+
+This purpose of this document is to describe our development workflow on
+GitHub; in particular, how we use milestones and issues. We recognize that our
+many subprojects may use variations of these ideas. This particular workflow is
+based on the practices from the JupyterLab repo
+(https://github.com/jupyterlab/jupyterlab) and are optional for Jupyter
+subprojects to follow.
+
+## Milestones
+
+We use Github milestones to track issues and pull requests related to particular
+releases of a subproject (https://github.com/jupyterlab/jupyterlab/milestones).
+
+* These milestones should be labeled with the version numbers of the releases
+  (`4.3`, `5.0`, etc.).
+* Subprojects should only have a few numbered milestones. Beyond one or two
+  future releases, we have found that detailed milestones become meaningless.
+* Subprojects should also have a `Backlog` milestone for issues and pull
+  requests that have not been assigned to a particular release milestone. This
+  also includes new feature requests ("wishlist") issues that we haven't
+  committed to implementing. Essentially, `Backlog` means "at some point in the
+  future, maybe."
+* Subprojects should have a `No Action` milestone to indicate issues or pull
+  requests that were closed with no action or work being performed. In general
+  we try to close issues/PRs with this milestone.
+* The only time an issue should have no milestone is early on in the triage
+  process when it still has the `status:Needs Info` label.
+
+## Labels
+
+We use GitHub labels to organize issues and pull requests
+(https://github.com/jupyterlab/jupyterlab/labels). The Github issues namespace
+is completely unstructured; each label is simply a name with an associated
+color. This unstructured nature, makes it difficult to organize work
+effectively. To address, this challenge, we are started to use a more formal
+label schema/taxonomy.
+
+First, every label should have one of four prefixes:
+
+1. `type:`
+2. `component:`
+3. `cat:`
+4. `status:`
+
+All issues of a given prefix should have the same color to make visual
```

Comment:
We find that this label scheme lends itself to a slightly different type of triage work. In the traditional flat namespace approach, I would hunt through the issues for a given label (such as "bug"). In that approach having different colors for each label was really helpful. The classic example is hunting for "red" bug labels.

In this more structured approach, triage is broken into two phases:

* Visually scanning to make sure all issues have the "right set of colors". Minimally, each issue should have one yellow label (type), a milestone and a few blue ones (component). In this scan the individual labels (content of the type and component) are not too important. Because we are visually encoding the type and component with color, it is really easy to scan visually. This type of scan can be done very quickly - I did a pass on all 200 issues in a single morning.
* Active work on a particular milestone+component. For this, we use queries rather than visual scanning as it allows us to quickly focus on a very small part of the code base.  Because the number of issues on a given milestone+component is small, visual scanning based on color is less relevant at this point.

My goal here isn't to convince you to adopt this approach, but rather explain why we made the choice we did and and how we use the system.